server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -375,6 +377,13 @@ public String getHeader(String key) {
         return value;
     }
 
+    /**
+     * Returns the persistent header for the given key or <code>null</code> if not present - persistent headers cannot be stashed
+     */
+    public Object getPersistent(String key) {
```

Comment:
Can existing data be migrated from where it is to persistent headers without breaking plugins?  If not, this prevents changes to security data that plugins read today

Reply:
The header that plugins primarily use is [this](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L215) to parse the user from the ThreadContext.

```
threadContext.putTransient(OPENDISTRO_SECURITY_USER_INFO_THREAD_CONTEXT, joiner.toString());
```

See example in Anomaly-Detection of [parsing this value](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/timeseries/util/ParseUtils.java#L432-L444)

```
public static User getUserContext(Client client) {
    String userStr = client.threadPool().getThreadContext().getTransient(ConfigConstants.OPENSEARCH_SECURITY_USER_INFO_THREAD_CONTEXT);
    logger.debug("Filtering result by " + userStr);
    return User.parse(userStr);
}
```

-----

With the change introduced in this PR I don't intend to migrate that header into a persistent header. I plan to use this to duplicate (at least until thread context stashing is fully deprecated and a better was for plugins to switch from user context <-> plugin context is developed) this [header](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/auth/BackendRegistry.java#L361-L364) so that it is always available. 

For Centralized Scheduled Job Identity Management using Identity this is the flow.

1.  Call on Create Detector endpoint
```
POST _plugins/_anomaly_detection/detectors
{
  "name": "test-detector",
  "description": "Test detector",
  "time_field": "timestamp",
  "indices": [
    "server_log*"
  ],
  ...
}
```

2. Anomaly Detection reads the user from the ThreadContext
3. Anomaly Detection stashes the ThreadContext
4. Anomaly Detection writes indexes a document to .opendistro-anomaly-detection-jobs with the job details
5. Job Scheduler listens to index operations on the .opendistro-anomaly-detection-jobs
6. In the preIndex hook, JobScheduler invokes IdentityPlugin -> Identity.associatedScheduledJobWithUser
7. The Security Plugin (implements IdentityPlugin) reads the User in the from the persistent ThreadContext header (remember that ThreadContext is stashed) and stores in the .opendistro-security-job-operators index

----

I think persistent ThreadContext headers solve this problem and could be used for the SecurityContext that has been brought up issues in the security repo.

Reply:
> I plan to use this to duplicate (at least until thread context stashing is fully deprecated and a better was for plugins to switch from user context <-> plugin context is developed)

If the data is duplicated and not used, there will be more latency for serialization and bandwidth to transmit this information, this doesn't seem like a good trade-off for the proposed future state.  If the intention is to replace how the plugin works, can we plan that out first so we know if this PR is necessary?

Reply:
It would be used and duplicated is the wrong word. The persistent header would always contain the authenticated user and not necessarily the current user which would be placed as a transient header. i.e. if a plugin stashes the threadcontext (meaning they switch from user context to plugin context) then the transient header could be erased or changed to a plugin user, but the persistent header would always be present.

This PR creates a generic means of putting headers into the ThreadContext that are unstashable.

@peternied Can you please chime in on the following issues:

- https://github.com/opensearch-project/security/pull/2861
- https://github.com/opensearch-project/security/issues/2909

There is pressure to make 2.9 and I need help with moving this along.